### PR TITLE
[SYCL][libdevice] Add tests for intel math device library.

### PR DESCRIPTION
**Pull request by jinge90**
Signed-off-by: jinge90 <ge.jin@intel.com>

### CODE REVIEW

**Comment by jinge90** [comment]
Hi, @aelovikov-intel 
It seems I still can't  add you to reviewer list. Could you add yourself to reviewers?
Thanks very much.

File: SYCL/DeviceLib/imf_fp64_test.cpp [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+          refer_vals[idx1][idx2]) {
+        std::cout << "__imf_double2float_* failed! idx1 = " << idx1
+                  << ", idx2 = " << idx2 << std::endl;
+        pass = false;
+        break;
+      }
+    }
+  }
+
+  assert(pass);
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_fp64_test.cpp]: nit: I'd move it to line 78 as `assert(false)`. That way it'd take one less step to hit the condition in the debugger (no breakpoint needed) and there will be no need for an extra variable `pass`.

File: SYCL/DeviceLib/imf_fp64_test.cpp [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+      {0, 0, 0, 0},
+      {0x40200000, 0x40200000, 0x40200000, 0x40200000},
+      {0x41bac425, 0x41bac425, 0x41bac426, 0x41bac425},
+      {0x51e6bd56, 0x51e6bd56, 0x51e6bd57, 0x51e6bd56},
+      {0xc2254cbd, 0xc2254cbd, 0xc2254cbc, 0xc2254cbc},
+      {0xea5f7b26, 0xea5f7b25, 0xea5f7b25, 0xea5f7b25},
+      {0x1fe141c2, 0x1fe141c3, 0x1fe141c3, 0x1fe141c2},
+      {0xdfd87e56, 0xdfd87e55, 0xdfd87e55, 0xdfd87e55}};
+  {
+    s::buffer<float, 2> output_buf(reinterpret_cast<float *>(output_vals),
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_fp64_test.cpp]:
```suggestion
    s::buffer<float, 2> output_buf(&output_vals[0][0],
```

File: SYCL/DeviceLib/imf_fp64_test.cpp [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+int __imf_double2loint(double);
+int __imf_double2int_rd(double);
+int __imf_double2int_rn(double);
+int __imf_double2int_ru(double);
+int __imf_double2int_rz(double);
+long long int __imf_double2ll_rd(double);
+long long int __imf_double2ll_rn(double);
+long long int __imf_double2ll_ru(double);
+long long int __imf_double2ll_rz(double);
+}
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_fp64_test.cpp]: I'm not sure how I feel about those. How would a user write the code to trigger execution of these implementations?

File: SYCL/DeviceLib/imf_fp64_test.cpp (hunk 1 above)
Comment (aelovikov-intel) [reply SYCL/DeviceLib/imf_fp64_test.cpp]: Same for the case below.

File: SYCL/DeviceLib/imf_fp64_test.cpp [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  float output_vals[NUM][4];
+  unsigned int refer_vals[NUM][4] = {
+      {0, 0, 0, 0},
+      {0x40200000, 0x40200000, 0x40200000, 0x40200000},
+      {0x41bac425, 0x41bac425, 0x41bac426, 0x41bac425},
+      {0x51e6bd56, 0x51e6bd56, 0x51e6bd57, 0x51e6bd56},
+      {0xc2254cbd, 0xc2254cbd, 0xc2254cbc, 0xc2254cbc},
+      {0xea5f7b26, 0xea5f7b25, 0xea5f7b25, 0xea5f7b25},
+      {0x1fe141c2, 0x1fe141c3, 0x1fe141c3, 0x1fe141c2},
+      {0xdfd87e56, 0xdfd87e55, 0xdfd87e55, 0xdfd87e55}};
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_fp64_test.cpp]: Do we have a native C++ reference implementation of those (i.e. `std::*`) ?

**Review by xtian-github — DISMISSED** [review]
LGTM

File: SYCL/DeviceLib/imf_fp64_test.cpp (hunk 4 above)
Comment (jinge90) [reply SYCL/DeviceLib/imf_fp64_test.cpp]: Hi, @aelovikov-intel 
There is no native C++ reference implementation for those APIs.

Thanks very much.

File: SYCL/DeviceLib/imf_fp32_test.cpp [hunk 5]
```diff
@@ -0,0 +1,1289 @@
...
+
+// RUN: %clangxx -fsycl -fno-builtin -fsycl-device-lib-jit-link %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+//
+// UNSUPPORTED: cuda || hip
+
+#include <cassert>
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_fp32_test.cpp]: I would suggest to remove unimportant code duplicate by doing something like this:

```
#include <sycl/sycl.hpp>

template <class InputTy, class OutputTy, class FuncTy>
void test(sycl::queue &q, std::initializer_list<InputTy> Input,
          std::initializer_list<OutputTy> RefOutput, FuncTy Func,
          int Line = __builtin_LINE()) {
  auto Size = Input.size();
  assert(RefOutput.size() == Size);

  sycl::buffer<InputTy> InBuf(Size);
  {
    sycl::host_accessor InAcc(InBuf, sycl::write_only);
    int i = 0;
    for (auto x : Input)
      InAcc[i++] = x;
  }

  sycl::buffer<OutputTy> OutBuf(Size);
  q.submit([&](sycl::handler &CGH) {
     sycl::accessor InAcc(InBuf, CGH, sycl::read_only);
     sycl::accessor OutAcc(OutBuf, CGH, sycl::write_only);
     CGH.parallel_for(Size,
                      [=](sycl::id<1> Id) { OutAcc[Id] = Func(InAcc[Id]); });
   }).wait();

  sycl::host_accessor Acc(OutBuf, sycl::read_only);
  for (int i = 0; i < Size; ++i) {
    auto Expected = *(std::begin(RefOutput) + i);
    if (Expected == Acc[i])
      continue;

    std::cout << "Mismatch at line " << Line << "[" << i << "]: " << Acc[i]
              << " != " << Expected << ", input was "
              << *(std::begin(Input) + i) << std::endl;
  }
}

int foo(float x) { return x + 1; }
int foo2(float x) { return x < 5.0 ? x + 1 : x; }

// Can't pass function pointer into kernel, wrap in a lambda.
#define F(Name) [](auto x) { return (Name)(x); }

int main() {
  // Default-ctor isn't SYCL2020 compatible yet.
  sycl::queue q{sycl::default_selector{}};
  test(q, {0.0f, 1.0f}, {1, 2}, F(foo));

  {
    std::initializer_list<float> Input = {1.0f, 3.0f, 6.0f};
    test(q, Input, {2, 4, 7}, F(foo));
    test(q, Input, {2, 4, 7}, F(foo2)); // Mismatch at line 52[2]: 6 != 7, input was 6
  }
}
```

Comment (aelovikov-intel) [reply SYCL/DeviceLib/imf_fp32_test.cpp]:
```
// In place of
std::initializer_list<float> Input = {1.0f, 3.0f, 6.0f};

// the following
auto Input = {1.0f, 3.0f, 6.0f};

// should also work.
```

Comment (aelovikov-intel) [reply SYCL/DeviceLib/imf_fp32_test.cpp]: Also, `F` can be changed into something like

```
// Maybe comma needs an extra step to properly expand
#define F(Name) [](auto x) { return (Name)(x); } , #Name
```

then the `test` could be modified to accept the function name as a parameter and report it on error as well.

File: SYCL/DeviceLib/imf_fp64_test.cpp [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+}
+
+int main(int, char **) {
+  s::default_selector device_selector;
+  s::queue device_queue(device_selector);
+  std::cout << "Running on "
+            << device_queue.get_device().get_info<s::info::device::name>()
+            << "\n";
+  bool is_host = device_queue.is_host();
+  if (is_host || device_queue.get_device().has(sycl::aspect::fp64)) {
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_fp64_test.cpp]: I'd suggest not testing it on the host *device* as it is removed in SYCL2020.

**Review by xtian-github — DISMISSED** [review]
LGTM

File: SYCL/DeviceLib/imf_utils.hpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  assert(Size == Input2.size());
+  assert(RefOutput.size() == Size);
+
+  sycl::buffer<InputTy> InBuf1(Size);
+  sycl::buffer<InputTy> InBuf2(Size);
+  {
+    sycl::host_accessor InAcc1(InBuf1, sycl::write_only);
+    sycl::host_accessor InAcc2(InBuf2, sycl::write_only);
+    int i = 0;
+    for (i = 0; i < Size; ++i) {
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_utils.hpp]:
```suggestion
    for (int i = 0; i < Size; ++i) {
```

File: SYCL/DeviceLib/imf_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+              << " != " << Expected << ", input was "
+              << *(std::begin(Input1) + i) << ", " << *(std::begin(Input2) + i)
+              << std::endl;
+    assert(false);
+  }
+}
+
+template <class InputTy1, class InputTy2, class InputTy3, class OutputTy,
+          class FuncTy>
+void test3(sycl::queue &q, std::initializer_list<InputTy1> Input1,
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_utils.hpp:89]: I don't like the duplication still, but we can address it later - might not be trivial.

File: SYCL/DeviceLib/imf_fp64_test.cpp [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+      test(device_queue, input_vals, ref_vals_ru,
+           FT(unsigned long long, __imf_ull2double_ru));
+      std::cout << "ull2double_ru passes." << std::endl;
+      test(device_queue, input_vals, ref_vals_rz,
+           FT(unsigned long long, __imf_ull2double_rz));
+      std::cout << "ull2double_rz passes." << std::endl;
+    }
+  } else {
+    std::cout << "Test skipped on platform without fp64 support." << std::endl;
+  }
```
Comment (aelovikov-intel) [inline SYCL/DeviceLib/imf_fp64_test.cpp]: Please use early return for this.

**Review by aelovikov-intel — DISMISSED** [review]

File: SYCL/DeviceLib/imf_utils.hpp (hunk 8 above)
Comment (jinge90) [reply SYCL/DeviceLib/imf_utils.hpp:89]: OK, let's commit the test now and I will address it later.

Thanks very much.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by jinge90** [comment]
Hi, @pvchupin 
Could you take a look at this PR, the pre-ci has 2 fails but not related to this PR.

[2022-08-20T03:15:53.959Z] 3 warning(s) in tests
[2022-08-20T03:15:53.959Z] ********************
[2022-08-20T03:15:53.959Z] Failed Tests (2):
[2022-08-20T03:15:53.959Z]   SYCL :: Plugin/level_zero_batch_event_status.cpp
[2022-08-20T03:15:53.959Z]   SYCL :: USM/depends_on.cpp
[2022-08-20T03:15:53.959Z] 
[2022-08-20T03:15:53.959Z] 
[2022-08-20T03:15:53.959Z] Testing Time: 1728.62s
[2022-08-20T03:15:53.959Z]   Unsupported      : 303
[2022-08-20T03:15:53.959Z]   Passed           : 624
[2022-08-20T03:15:53.959Z]   Expectedly Failed:  11
[2022-08-20T03:15:53.959Z]   Failed           :   2

Thanks very much.

**Comment by aelovikov-intel** [comment]
Failures look known, @intel/llvm-gatekeepers can you please merge this in?

**Pull request merged**